json/parse_fuzzer.cc

```diff
@@ -0,0 +1,33 @@
+// Copyright 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <iostream>
+#include <sstream>
+#include <json.hpp>
+
+using json = nlohmann::json;
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+  try {
+    std::stringstream s;
+    s << json::parse(data, data + size);
+    try {
+      auto j2 = json::parse(s.str());
```

Comment:
This technique finds tons of bugs and should be used for all libraries that has round-trip, i.e. serialization/deserialization, compression/decompression, conversion to/from, etc. Bugs found by such checks (logical bugs) are often valued by developers more than crashes on insane data.

I would also suggest to compare results of first and second parse. That catches losing, doubling, altering (e.g. double/missed escape/unescape) data during round-trip.